ophyd/controls/device.py

```diff
@@ -0,0 +1,439 @@
+        '''Wait for signals to connect
+
+        Parameters
+        ----------
+        all_signals : bool, optional
+            Wait for all signals to connect (including lazy ones)
+        timeout : float or None
+            Overall timeout
+        '''
+        names = [attr for cpt, attr in self._sig_attrs.items()
+                 if not cpt.lazy or all_signals]
+
+        # Instantiate first to kickoff connection process
+        signals = [getattr(self, name) for name in names]
+
+        t0 = time.time()
+        while timeout is None or (time.time() - t0) < timeout:
+            connected = [sig.connected for sig in signals]
+            if all(connected):
+                return
+            time.sleep(min((0.05, timeout / 10.0)))
+
+        unconnected = [sig.name for sig in signals
+                       if not sig.connected]
+
+        raise TimeoutError('Failed to connect to all signals: {}'
+                           ''.format(', '.join(unconnected)))
+
+    @property
+    def connected(self):
+        return all(signal.connected for name, signal in self._signals.items())
+
+    def _get_devattr(self, name):
+        '''Gets a device attribute which may come from a sub-device'''
+        try:
+            owner_attr = self._sig_owner[name]
+        except KeyError:
+            raise ValueError('Unknown read signal: {}'.format(name))
+
+        if owner_attr is None:
+            # None means this instance owns it
+            return getattr(self, name)
+        else:
+            # Otherwise, get the owner first
+            owner = getattr(self, owner_attr)
+            # Then the attribute
+            return getattr(owner, name)
+
+    def read(self):
+        # map names ("data keys") to actual values
+        values = OrderedDict()
+        for name in self.read_signals:
+            signal = self._get_devattr(name)
+            values.update(signal.read())
+
+        return values
+
+    def describe(self):
+        desc = OrderedDict()
+        for name in self.read_signals:
+            signal = self._get_devattr(name)
+            desc.update(signal.describe())
+
+        return desc
+
+    @property
+    def trigger_signals(self):
+        names = [attr for cpt, attr in self._sig_attrs.items()
+                 if cpt.trigger_value is not None]
+
+        return [getattr(self, name) for name in names]
+
+    def trigger(self, **kwargs):
+        """Start acquisition"""
+        # TODO mass confusion here
+        signals = self.trigger_signals
+        if len(signals) > 1:
+            raise NotImplementedError('TODO more than one trigger')
+        elif len(signals) == 0:
+            raise RuntimeError('Device has no trigger signal(s)')
+
+        acq_signal = signals[0]
+        status = DeviceStatus(self)
+        self.subscribe(status._finished,
+                       event_type=self.SUB_ACQ_DONE, run=False)
+
+        def done_acquisition(**kwargs):
```

Comment:
kwargs are necessary because of the EpicsSignal subscription callback signature - and ignored as the contents are unimportant. The important part is that the callback itself was called to indicate it has completed.